include/fast_float/ascii_number.h

```diff
@@ -47,6 +64,40 @@ uint64_t read_u64(const char *chars) {
   return val;
 }
 
+#ifdef FASTFLOAT_SSE2
+
+fastfloat_really_inline
+uint64_t simd_read8_to_u64(const __m128i data) {
+FASTFLOAT_SIMD_DISABLE_WARNINGS
+  static const char16_t kmasks[] = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff };
+  const __m128i masks = _mm_loadu_si128(reinterpret_cast<const __m128i*>(kmasks));
+
+  // todo: with AVX512BW + AVX512VL, can use cvtepi16_epi8 instead of masking + pack
+  __m128i masked = _mm_and_si128(data, masks);
+  __m128i packed = _mm_packus_epi16(masked, masked);
+
+  uint64_t val;
```

Comment:
Possibly the store_u could be replace by _mm_cvtsi128_si64x which may not be simpler but I don't think that's even right.

You already have the value in a SIMD register... finish the job right away... something like this...

```
const __m128i ascii0 = _mm_set1_epi8('0');
  const __m128i mul_1_10 =
      _mm_setr_epi8(10, 1, 10, 1, 10, 1, 10, 1, 10, 1, 10, 1, 10, 1, 10, 1);
  const __m128i mul_1_100 = _mm_setr_epi16(100, 1, 100, 1, 100, 1, 100, 1);
  const __m128i mul_1_10000 =
      _mm_setr_epi16(10000, 1, 10000, 1, 10000, 1, 10000, 1);
  const __m128i input = _mm_sub_epi8(
      _mm_loadu_si128(reinterpret_cast<const __m128i *>(chars)), ascii0);
  const __m128i t1 = _mm_maddubs_epi16(input, mul_1_10);
  const __m128i t2 = _mm_madd_epi16(t1, mul_1_100);
  const __m128i t3 = _mm_packus_epi32(t2, t2);
  const __m128i t4 = _mm_madd_epi16(t3, mul_1_10000);
  return _mm_cvtsi128_si32(
      t4); 
```

Well. It is worth investigating.

Reply:
I had something similar before in this PR (which I took from http://0x80.pl/articles/simd-parsing-int-sequences.html and modified for SSE2) but I removed it because the improvement wasn't much and I thought it might be difficult to maintain.

Here's a run on WSL2. I see an improvement of 3-4% over multiple runs. 
```
# read 111126 lines
volume = 2.03972 MB
char_type = char
netlib                                  :   270.02 MB/s (+/- 3.0 %)    14.71 Mfloat/s      29.59 i/B   569.44 i/f (+/- 0.0 %)     13.91 c/B   267.69 c/f (+/- 1.5 %)      2.13 i/c      3.94 GHz
doubleconversion                        :   211.86 MB/s (+/- 4.3 %)    11.54 Mfloat/s      55.18 i/B  1061.94 i/f (+/- 0.0 %)     17.70 c/B   340.69 c/f (+/- 3.3 %)      3.12 i/c      3.93 GHz
strtod                                  :   143.99 MB/s (+/- 2.3 %)     7.84 Mfloat/s      66.12 i/B  1272.61 i/f (+/- 0.0 %)     26.02 c/B   500.86 c/f (+/- 1.2 %)      2.54 i/c      3.93 GHz
abseil                                  :   342.49 MB/s (+/- 4.2 %)    18.66 Mfloat/s      29.60 i/B   569.73 i/f (+/- 0.0 %)     10.98 c/B   211.31 c/f (+/- 2.6 %)      2.70 i/c      3.94 GHz
fastfloat                               :   756.29 MB/s (+/- 8.6 %)    41.20 Mfloat/s      16.75 i/B   322.46 i/f (+/- 0.0 %)      4.98 c/B    95.85 c/f (+/- 6.5 %)      3.36 i/c      3.95 GHz
fastfloat_simd                          :   700.84 MB/s (+/- 9.3 %)    38.18 Mfloat/s      16.75 i/B   322.45 i/f (+/- 0.0 %)      5.37 c/B   103.29 c/f (+/- 7.5 %)      3.12 i/c      3.94 GHz
fastfloat_simd2                         :   700.00 MB/s (+/- 8.9 %)    38.14 Mfloat/s      16.75 i/B   322.45 i/f (+/- 0.0 %)      5.37 c/B   103.37 c/f (+/- 6.3 %)      3.12 i/c      3.94 GHz


volume = 4.07944 MB
char_type = char16_t
doubleconversion                        :   414.08 MB/s (+/- 5.4 %)    11.28 Mfloat/s      27.56 i/B  1060.95 i/f (+/- 0.0 %)      9.27 c/B   356.66 c/f (+/- 2.2 %)      2.97 i/c      4.02 GHz
fastfloat                               :  1180.94 MB/s (+/- 9.1 %)    32.17 Mfloat/s      10.20 i/B   392.51 i/f (+/- 0.0 %)      3.24 c/B   124.72 c/f (+/- 6.1 %)      3.15 i/c      4.01 GHz
fastfloat_simd                          :  1290.11 MB/s (+/- 8.2 %)    35.14 Mfloat/s      10.20 i/B   392.65 i/f (+/- 0.0 %)      2.90 c/B   111.82 c/f (+/- 7.1 %)      3.51 i/c      3.93 GHz
fastfloat_simd2                         :  1330.85 MB/s (+/- 6.9 %)    36.25 Mfloat/s      10.23 i/B   393.63 i/f (+/- 0.0 %)      2.82 c/B   108.58 c/f (+/- 6.0 %)      3.63 i/c      3.94 GHz
```

If we include this then we should also add workflows to test it (since it is SSE4.1 and compilers only make SSE2 available by default).

Reply:
You are right that we don't expect a large gain, and the fact that it is not basic SSE2 is a serious concern because we have, thus far, stayed away from runtime dispatching in this library. We should only throw more sophistication at the problem if it is warranted by significant gains.

Reply:
GCC and clang define the macro \_\_SSE_4_1\_\_ when SSE4.1 is available, so it should just be a matter of checking if it exists and adding the -msse4.1 cmdline option to the workflows (-msse4.1 is required regardless of target). MSVC doesn't provide this macro but we can check for \_\_AVX\_\_ which should imply SSE4.1.  
With the corrected math, I see only a 2-3% improvement though. Should I add it?